NoWhitespaceAfter synchronized method

Hello,

I have the following checkstyle configuration in `pom.xml`:
```xml
<module name="NoWhitespaceAfter">
      <property name="tokens" value="ARRAY_INIT,AT,INC,DEC,UNARY_MINUS,UNARY_PLUS,BNOT,LNOT,DOT,TYPECAST,ARRAY_DECLARATOR,INDEX_OP,LITERAL_SYNCHRONIZED,METHOD_REF" />
</module>
```

With a `mvn verify`, the following piece of code reports an error `'synchronized' is followed by whitespace`:

```java
public class TestClass {
    private synchronized Row waitForNextRow() {
    }
}
```

It seems to be a false positive because the check should only apply on structures like:
```java
synchronized(lock) { ... }
```

Here is the output:
```
$ cat TestClass.java
public class TestClass {
    private synchronized Row waitForNextRow() {
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="NoWhitespaceAfter">
      <property name="tokens" value="ARRAY_INIT,AT,INC,DEC,UNARY_MINUS,UNARY_PLUS,BNOT,LNOT,DOT,TYPECAST,ARRAY_DECLARATOR,INDEX_OP,LITERAL_SYNCHRONIZED,METHOD_REF" />
</module>
    </module>
</module>

$ java -jar checkstyle-10.5.0-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:2:13: 'synchronized' is followed by whitespace. [NoWhitespaceAfter]
Audit done.
Checkstyle ends with 1 errors.
```

Thanks & best regards.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java
@@ -283,12 +283,8 @@ public class NoWhitespaceAfterCheck extends AbstractCheck {
      * @return true if whitespace after ast should be checked
      */
     private static boolean shouldCheckWhitespaceAfter(DetailAST ast) {
-        boolean checkWhitespace = true;
         final DetailAST previousSibling = ast.getPreviousSibling();
-        if (previousSibling != null && previousSibling.getType() == TokenTypes.ANNOTATIONS) {
-            checkWhitespace = false;
-        }
-        return checkWhitespace;
+        return previousSibling == null || previousSibling.getType() != TokenTypes.ANNOTATIONS;
     }
 
     /**

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java
@@ -55,6 +55,13 @@ import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
  * <pre>
  * public void foo(final char @NotNull [] param) {} // No violation
  * </pre>
+ * <p>
+ * Note: This check processes the
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_SYNCHRONIZED">
+ * LITERAL_SYNCHRONIZED</a> token only when it appears as a part of a
+ * <a href="https://docs.oracle.com/javase/specs/jls/se19/html/jls-14.html#jls-14.19">
+ * synchronized statement</a>, i.e. {@code synchronized(this) {}}.
+ * </p>
  * <ul>
  * <li>
  * Property {@code allowLineBreaks} - Control whether whitespace is allowed
@@ -284,7 +291,10 @@ public class NoWhitespaceAfterCheck extends AbstractCheck {
      */
     private static boolean shouldCheckWhitespaceAfter(DetailAST ast) {
         final DetailAST previousSibling = ast.getPreviousSibling();
-        return previousSibling == null || previousSibling.getType() != TokenTypes.ANNOTATIONS;
+        final boolean isSynchronizedMethod = ast.getType() == TokenTypes.LITERAL_SYNCHRONIZED
+                        && ast.getFirstChild() == null;
+        return !isSynchronizedMethod
+                && (previousSibling == null || previousSibling.getType() != TokenTypes.ANNOTATIONS);
     }
 
     /**

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java
@@ -50,11 +50,9 @@ import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
  * INDEX_OP</a> will be ignored.
  * </p>
  * <p>
- * If the annotation is between the type and the array, the check will skip validation for spaces
+ * If the annotation is between the type and the array, like {@code char @NotNull [] param},
+ * the check will skip validation for spaces.
  * </p>
- * <pre>
- * public void foo(final char @NotNull [] param) {} // No violation
- * </pre>
  * <p>
  * Note: This check processes the
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_SYNCHRONIZED">

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/NoWhitespaceAfterCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/NoWhitespaceAfterCheck.xml
@@ -30,7 +30,14 @@
  &lt;/p&gt;
  &lt;pre&gt;
  public void foo(final char @NotNull [] param) {} // No violation
- &lt;/pre&gt;</description>
+ &lt;/pre&gt;
+ &lt;p&gt;
+ Note: This check processes the
+ &lt;a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_SYNCHRONIZED"&gt;
+ LITERAL_SYNCHRONIZED&lt;/a&gt; token only when it appears as a part of a
+ &lt;a href="https://docs.oracle.com/javase/specs/jls/se19/html/jls-14.html#jls-14.19"&gt;
+ synchronized statement&lt;/a&gt;, i.e. {@code synchronized(this) {}}.
+ &lt;/p&gt;</description>
          <properties>
             <property default-value="true" name="allowLineBreaks" type="boolean">
                <description>Control whether whitespace is allowed

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/NoWhitespaceAfterCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/NoWhitespaceAfterCheck.xml
@@ -26,11 +26,9 @@
  INDEX_OP&lt;/a&gt; will be ignored.
  &lt;/p&gt;
  &lt;p&gt;
- If the annotation is between the type and the array, the check will skip validation for spaces
+ If the annotation is between the type and the array, like {@code char @NotNull [] param},
+ the check will skip validation for spaces.
  &lt;/p&gt;
- &lt;pre&gt;
- public void foo(final char @NotNull [] param) {} // No violation
- &lt;/pre&gt;
  &lt;p&gt;
  Note: This check processes the
  &lt;a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_SYNCHRONIZED"&gt;

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java
@@ -354,6 +354,17 @@ public class NoWhitespaceAfterCheckTest
                 getPath("InputNoWhitespaceAfterWithEmoji.java"), expected);
     }
 
+    @Test
+    public void testNoWhitespaceAfterSynchronized() throws Exception {
+        final String[] expected = {
+            "18:9: " + getCheckMessage(MSG_KEY, "synchronized"),
+        };
+
+        verifyWithInlineConfigParser(
+                getPath("InputNoWhitespaceAfterSynchronized.java"),
+                expected);
+    }
+
     /**
      * Creates MOCK lexical token and returns AST node for this token.
      *

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterSynchronized.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterSynchronized.java
@@ -1,0 +1,59 @@
+/*
+NoWhitespaceAfter
+allowLineBreaks = (default)true
+tokens = LITERAL_SYNCHRONIZED
+
+
+*/
+package com.puppycrawl.tools.checkstyle.checks.whitespace.nowhitespaceafter;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Target;
+
+public class InputNoWhitespaceAfterSynchronized<T> {
+
+    private final Object lock = new Object();
+
+    void m1() {
+        synchronized (lock) {} // violation
+        synchronized(lock) {} // ok
+    }
+
+    synchronized int m2() { // ok
+        return 2;
+    }
+
+    synchronized private int[] m3() { // ok
+        return new int[]{2};
+    }
+
+    synchronized int[] m4() { // ok
+        return new int[]{2};
+    }
+
+    private final synchronized int m5() { // ok
+        return 2;
+    }
+
+    synchronized T m6() { // ok
+        return null;
+    }
+
+    @SyncAnno synchronized T m7() { // ok
+        return null;
+    }
+
+    @SyncAnno private synchronized T m8() { // ok
+        return null;
+    }
+
+    // Note that we do not check 'synchronized' as a modifier,
+    // convention is to have type parameters of a generic method
+    // be surrounded by whitespace.
+    synchronized <T2> void m9() {} // ok
+    synchronized<T2> void m10() {} // ok
+}
+
+@Target(ElementType.METHOD)
+@interface SyncAnno {
+}

--- a/src/xdocs/config_whitespace.xml
+++ b/src/xdocs/config_whitespace.xml
@@ -1347,12 +1347,10 @@ import static java.math.BigInteger.ZERO;
           will be ignored.
         </p>
         <p>
-         If the annotation is between the type and the array, the check will skip validation for
-         spaces
+          If the annotation is between the type and the array,
+          like <code>char @NotNull [] param</code>, the check will skip
+          validation for spaces.
         </p>
-        <source>
-public void foo(final char @NotNull [] param) {} // No violation
-        </source>
         <p>
          Note: This check processes the
           <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_SYNCHRONIZED">

--- a/src/xdocs/config_whitespace.xml
+++ b/src/xdocs/config_whitespace.xml
@@ -1353,6 +1353,13 @@ import static java.math.BigInteger.ZERO;
         <source>
 public void foo(final char @NotNull [] param) {} // No violation
         </source>
+        <p>
+         Note: This check processes the
+          <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_SYNCHRONIZED">
+         LITERAL_SYNCHRONIZED</a> token only when it appears as a part of a
+          <a href="https://docs.oracle.com/javase/specs/jls/se19/html/jls-14.html#jls-14.19">
+            synchronized statement</a>, i.e. <code>synchronized(this) {}</code>.
+        </p>
       </subsection>
 
       <subsection name="Properties" id="NoWhitespaceAfter_Properties">

